### PR TITLE
Add player history

### PR DESCRIPTION
**Pull request by RipleyTom**
Implement all APIs related to player history and shows a list of it in rpcn=>friends where you can add friends quickly that way. Of note is that this list is purely local to your instance of RPCS3 and not stored on RPCN.

### CODE REVIEW

File: rpcs3/Emu/NP/np_handler.h [hunk 1]
```diff
@@ -40,6 +40,13 @@ namespace np
 		} data;
 	};
 
+	struct player_history
+	{
+		u64 timestamp;
```
Comment (Megamouse) [inline rpcs3/Emu/NP/np_handler.h:45]:
```suggestion
		u64 timestamp{};
```

File: rpcs3/Emu/NP/np_handler.cpp [hunk 2]
```diff
@@ -1217,13 +1282,169 @@ namespace np
...
+
+		if (all_history)
+		{
+			return ::size32(players_history);
+		}
+
+		const std::string communication_id_str = std::string(basic_handler.context.data);
+		u32 count = 0;
+
+		for (auto it = players_history.begin(); it != players_history.end(); it++)
```
Comment (Megamouse) [inline rpcs3/Emu/NP/np_handler.cpp:1363]: Std::count_if ?

File: rpcs3/rpcs3qt/rpcn_settings_dialog.cpp [hunk 3]
```diff
@@ -858,6 +858,19 @@ void friend_callback(void* param, rpcn::NotificationType ntype, const std::strin
 	dlg->callback_handler(ntype, username, status);
 }
 
+// Avoid including np_handler.h
+namespace np
+{
+	struct player_history
+	{
+		u64 timestamp;
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/rpcn_settings_dialog.cpp:866]:
```suggestion
		u64 timestamp{};
```

File: rpcs3/rpcs3qt/rpcn_settings_dialog.cpp [hunk 4]
```diff
@@ -1014,6 +1049,38 @@ rpcn_friends_dialog::rpcn_friends_dialog(QWidget* parent)
...
+			QMenu* context_menu = new QMenu();
+			QAction* send_friend_request_action = context_menu->addAction(tr("&Send Friend Request"));
+
+			connect(send_friend_request_action, &QAction::triggered, this, [this, str_sel_friend]()
+				{
+					if (!m_rpcn->add_friend(str_sel_friend))
+					{
+						QMessageBox::critical(this, tr("Error sending a friend request!"), tr("An error occurred while trying to send a friend request!"), QMessageBox::Ok);
+					}
+					else
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/rpcn_settings_dialog.cpp:1072]: Early out instead of else

File: rpcs3/rpcs3qt/rpcn_settings_dialog.cpp [hunk 5]
```diff
@@ -945,6 +966,20 @@ rpcn_friends_dialog::rpcn_friends_dialog(QWidget* parent)
 		add_update_list(m_lst_blocks, QString::fromStdString(blck), m_red_icon, QVariant(false));
 	}
 
+	auto history = np::load_players_history();
+	std::map<u64, std::string, std::greater<u64>> sorted_history;
+
+	for (const auto& [username, user_info] : history)
+	{
+		if (!data.friends.contains(username) && !data.requests_sent.contains(username) && !data.requests_received.contains(username))
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/rpcn_settings_dialog.cpp:974]: I'm a bit confused about this.
So this list only shows players that you played but doesn't list the players you know ?

File: rpcs3/rpcs3qt/rpcn_settings_dialog.cpp [hunk 6]
```diff
@@ -945,6 +966,20 @@ rpcn_friends_dialog::rpcn_friends_dialog(QWidget* parent)
...
+			sorted_history.insert(std::make_pair(user_info.timestamp, std::move(username)));
+	}
+
+	for (const auto& [_, username] : sorted_history)
+	{
+		m_lst_history->addItem(new QListWidgetItem(QString::fromStdString(username)));
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/rpcn_settings_dialog.cpp:980]: you could add the description as tooltext here.

File: rpcs3/rpcs3qt/rpcn_settings_dialog.cpp (hunk 5 above)
Comment (RipleyTom) [reply rpcs3/rpcs3qt/rpcn_settings_dialog.cpp:974]: Yes, the idea is to have the list of players you recently played with and you're not already friends with.

**Pull request merged**